Song list page added

WangRemy opened the pull request.

WangRemy committed:
WIPTEST

WangRemy committed:
Added Song List page

Used generic ListView class to display all the songs in a table of
the Song model.

SanderBeekhuis commented on song_rater/admin.py:
```diff
@@ -1,3 +1,6 @@
 from django.contrib import admin
 
 # Register your models here.
+
+from.models import Song
```
Can you drop this comment and neatly organize your imports?

SanderBeekhuis commented on song_rater/models/song.py:
```diff
@@ -1,4 +1,8 @@
+    title = models.CharField(max_length = 128)
+    author = models.CharField(max_length = 64)
+
+    def __str__(self):
+        return f"{self.title} - {self.author}"
```
Neat that you added an `__str__` representation.

SanderBeekhuis commented on templates/song_rater/song_list.html:
```diff
@@ -0,0 +1,24 @@
+    <title>Song_List</title>
+</head>
+<body>
+<h1>List of all Songs</h1>
+<table border="1" width="600">
+<tr>
+<td>Title</td>
+<td>Author</td>
+</tr>
+
+{% for song in object_list%}
+<tr>
+<td>{{ song.title }}</td>
+<td>{{ song.author }}</td>
+</tr>
+{% endfor %}
+</table>
+</body>
+</html>
+
```
Try indenting your HTML to increase readability. :)

SanderBeekhuis reviewed (changes requested):
Overall very neat code Remy! I have  few small comments, those should be easy to resolve. 

I however miss test cases. Can you please add some test cases?

For example?
- I can navigate to the song list
- And, Given a few songs in the database, I see their names in the song list page.

These test might seem silly now, but it's best to start writing tests early. This way it will be easy to expand on these test when we add "average ratings" etc.

SebastiaanZ commented on templates/song_rater/song_list.html:
```diff
@@ -0,0 +1,24 @@
+<td>{{ song.title }}</td>
+<td>{{ song.author }}</td>
+</tr>
+{% endfor %}
```
Wouldn't it be nice to display something here in case there are no songs in the list yet? Django has a special [`empty`](https://docs.djangoproject.com/en/3.1/ref/templates/builtins/#for-empty) template tag that you could use for that situation. It's a part of the `for`-loop.

SebastiaanZ commented on templates/song_rater/song_list.html:
```diff
@@ -0,0 +1,24 @@
+<td>Author</td>
+</tr>
+
+{% for song in object_list%}
```
To increase the readability of this for loop, you could set a `context_object_name` attribute on your `ListView` class to set a custom name for the `object_list` context variable. I think that something like `song_list` would be nice. It doesn't really matter in this instance, as it's small, straightforward template, but it could help in more complicated situations.

SebastiaanZ commented on song_rater/views.py:
```diff
@@ -1,11 +1,16 @@
     return render(request, 'song_rater/index.html', context)
+
+
+class SongListView(ListView):
+    model = Song
```
This is where you could add a [`context_object_name`](https://docs.djangoproject.com/en/3.1/topics/class-based-views/generic-display/#making-friendly-template-contexts) attribute, as noted below in the comment attached to the template file.

WangRemy committed:
Cleaned up code

WangRemy committed:
Cleaned up html format

Changed object_list to song_list for readability

WangRemy committed:
Added testcase for Song_list page.

It checks if you can reach the page, did not yet
add the ability to check for songs on the page.
Will discuss this on 2/17/2021 at the stand-up.

WangRemy committed:
Added context object name for better readability.

SanderBeekhuis commented on song_rater/views.py:
```diff
@@ -1,4 +1,4 @@
-from django.shortcuts import render
+from django.shortcuts import render, get_object_or_404
```
You don't use this import

SanderBeekhuis commented on song_rater/tests.py:
```diff
@@ -1,6 +1,7 @@
-from django.test import TestCase
+from django.test import TestCase, Client
```
You don't use this import

SanderBeekhuis commented on song_rater/tests.py:
```diff
@@ -9,3 +10,12 @@ def test_index(self):
         self.assertContains(r, "Ordina Music Rater")
+
+class SongListTest(TestCase):
+    def test_no_songs(self):
```
Nice test! 

Let's add one (together) for when songs are present.

SanderBeekhuis reviewed (commented):
Getting there.

WangRemy committed:
Added testcase for view when songs are present

Processed review points.
